akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/HostConnectionPool.scala

```diff
@@ -0,0 +1,335 @@
+          slots.exists(_.isIdle)
+
+        def dispatchResponse(req: RequestContext, res: HttpResponse): Unit =
+          dispatchResponseContext(ResponseContext(req, Success(res)))
+
+        def dispatchFailure(req: RequestContext, cause: Throwable): Unit = {
+          // FIXME: or retry
+          if (req.retriesLeft > 0) {
+            log.debug("Request has {} retries left, retrying...", req)
+            retryBuffer.addLast(req.copy(retriesLeft = req.retriesLeft - 1))
+          } else
+            dispatchResponseContext(ResponseContext(req, Failure(cause)))
+        }
+
+        def dispatchResponseContext(resCtx: ResponseContext): Unit =
+          if (outBuffer.isEmpty && isAvailable(responsesOut))
+            push(responsesOut, resCtx)
+          else
+            outBuffer.addLast(resCtx)
+
+        def dispatchRequest(req: RequestContext): Unit = {
+          val slot =
+            slots.find(_.isIdle)
+              .getOrElse(throw new IllegalStateException("Tried to dispatch request when no slot is idle"))
+
+          slot.debug("Dispatching request") // FIXME: add abbreviation
+          slot.dispatchRequest(req)
+        }
+
+        private class Slot(val slotId: Int) extends SlotContext {
+          private[this] var state: SlotState = SlotState.Unconnected
+          private[this] var connection: SlotConnection = _
+          def isIdle: Boolean = state.isIdle
+          def isConnected: Boolean = state.isConnected
+          def shutdown(): Unit = {
+            // TODO: should we offer errors to the connection?
+            closeConnection()
+
+            state.onShutdown(this)
+          }
+
+          def initialize(): Unit =
+            if (slotId < settings.minConnections)
+              updateState(_.onPreConnect(this))
+
+          def onConnected(outgoing: Http.OutgoingConnection): Unit =
+            updateState(_.onConnectedAttemptSucceeded(this, outgoing))
+
+          def onConnectFailed(cause: Throwable): Unit =
+            updateState(_.onConnectionAttemptFailed(this, cause))
+
+          def dispatchRequest(req: RequestContext): Unit =
+            updateState(_.onNewRequest(this, req))
+
+          def onRequestEntityCompleted(): Unit =
+            updateState(_.onRequestEntityCompleted(this))
+          def onRequestEntityFailed(cause: Throwable): Unit =
+            updateState(_.onRequestEntityFailed(this, cause))
+
+          def onResponseReceived(response: HttpResponse): Unit =
+            updateState(_.onResponseReceived(this, response))
+
+          def onResponseEntityCompleted(): Unit =
+            updateState(_.onResponseEntityCompleted(this))
+          def onResponseEntityFailed(cause: Throwable): Unit =
+            updateState(_.onResponseEntityFailed(this, cause))
+
+          def onConnectionCompleted(): Unit = updateState(_.onConnectionCompleted(this))
+          def onConnectionFailed(cause: Throwable): Unit = updateState(_.onConnectionFailed(this, cause))
+
```

Comment:
this callback style reads excellent each time I see it in action ;)

Reply:
Yep, a bit verbose but it for reading the explicit style is quite good.